File: src/udp/udp.cpp
#include "../controller.h"
#include "../coroutine.h"
#include "udp.h"
#include "socket.h"
#include "server.h"

namespace flame::udp {

	std::unique_ptr<boost::asio::ip::udp::resolver> resolver_;

	void declare(php::extension_entry& ext) {
		gcontroller->on_init([] (const php::array& options) {
            resolver_.reset(new boost::asio::ip::udp::resolver(gcontroller->context_x));
        })->on_stop([] () {
            resolver_.reset();
        });
		ext
			.function<connect>("flame\\udp\\connect", {
				{"address", php::TYPE::STRING},
			});
		socket::declare(ext);
		server::declare(ext);
	}

	php::value connect(php::parameters& params) {
		php::object obj(php::class_entry<socket>::entry());
        socket *ptr = static_cast<socket *>(php::native(obj));

        std::string str = params[0];
        auto pair = addr2pair(str);
        if(pair.first.empty() || pair.second.empty()) throw php::exception(zend_ce_type_error
            , "Failed to connect UDP socket: illegal address format"
            , -1);

		coroutine_handler ch{coroutine::current};
        boost::system::error_code err;
		boost::asio::ip::udp::resolver::results_type eps;
		resolver_->async_resolve(pair.first, pair.second
            , [&err, &obj, &ptr, &eps, &ch] (const boost::system::error_code& error, boost::asio::ip::udp::resolver::results_type results) { // DNS 地址解析

			if(error) err = error;
			else eps = results;
			ch.resume();
		});
		ch.suspend();
		if(err) throw php::exception(zend_ce_exception
                , (boost::format("Failed to resolve address: %s") % err.message()).str()
                , err.value());

        boost::asio::async_connect(ptr->socket_, eps
            , [&err, &obj, &ptr, &ch](const boost::system::error_code &error, const boost::asio::ip::udp::endpoint &edp) { // 连接

            if (error) err = error;
			else {
            	obj.set("local_address", (boost::format("%s:%d") % ptr->socket_.local_endpoint().address().to_string() % ptr->socket_.local_endpoint().port()).str());
				obj.set("remote_address", (boost::format("%s:%d") % ptr->socket_.remote_endpoint().address().to_string() % ptr->socket_.remote_endpoint().port()).str());
			}
            ch.resume();
        });
        ch.suspend();
        if(err) throw php::exception(zend_ce_exception
            , (boost::format("Failed to connect UDP socket: %s") % err.message()).str()
            , err.value());
		ptr->connected_ = true;
        return std::move(obj);
	}
    
    std::pair<std::string, std::string> addr2pair(const std::string &addr) {
        char *s = const_cast<char*>(addr.data()), *p, *e = s + addr.size();
        for (p = e - 1; p > s; --p) if (*p == ':') break; // 分离 地址与端口
        if (*p != ':') return std::make_pair<std::string, std::string>(std::string(addr), "");
        else return std::make_pair<std::string, std::string>(
                std::string(s, p - s), std::string(p+1, e-p-1));
    }

} // namespace flame::tcp

File: doc/udp.php
<?php
/**
 * 提供对 UDP 网络操作的简单封装
 */
namespace flame\udp;

/**
 * 连接指定地址, 并返回 socket 对象
 * @param string $address 连接地址, 例如: "127.0.0.1:8687"
 * 注意: UDP "链接" 能在一定程度上提高效率;
 */
function connect(string $address):socket {}

class socket {
    /**
     * 构建 socket 可选的进行地址绑定
     * 实际函数拥有两种形式:
     *  __construct(string $address, array $option = [])
     * 或
     *  __construct(array $option = [])
     * 前者可用于地址绑定;
     * @param array $option 可用选项如下:
     * * `max` integer - 接收缓冲区上限(默认为 64k 可减小已提升效率)
     */
    function __construct(mixed $opt_or_address = null, array $option = []) {}
    /**
     * 本地地址（含端口）
     * @property string
     */
    public $local_address;
    /**
     * 远端地址（含端口）
     * @property string
     */
    public $remote_address;
    /**
     * 发送一个 UDP 包给已连接的对端
     * 注意: 只能在"连接"形态下使用
     * @see flame\udp\connect()
     */
    function send(string $data) {}
    /**
     * 向指定目标 $to 发送一个 UDP 包
     * @param string $to 目标地址, 可以是域名或地址;
     */
    function send_to(string $data, string $to) {}
    /**
     * 接收并返回一个 UDP 包
     * @return mixed 返回接收到的数据包(文本 string); 若 socket 被关闭则返回 null;
     */
    function recv():mixed {}
    /**
     * 关闭连接
     * 注意: 正在进行的发送 send(_to) 或 recv(_from) 接收动作将被终止;
     */
    function close() {}
}
/**
 * 简化的服务端, 相较使用上述 socket 自行实现效率稍高
 */
class server {
    /**
     * 创建一个服务器, 绑定在指定的地址上, 准备服务
     * @param string $address 绑定地址, 例如: "0.0.0.0:8888" 或 ":::8888"
     * @param string
     */
    function __construct(string $address, array $options = []) {}
    /**
     * 启动服务器, 当有连接建立时, 启用协程执行回调函数;
     * @param callable $cb 回调函数, 形如:
     *  function(string $data, string $from) {}
     *  $data - 接收的数据包
     *  $from - 来源地址
     */
    function run(callable $cb) {}
    /**
     * 使用当前服务器对应 UDP 套接字, 向指定地址发送消息
     * @see flame::socket::send_to()
     */
    function send_to(string $data, string $address) {}
    /**
     * 关闭服务器 (结束 run() 过程, 保证正在进行处理的"消费"协程完成)
     */
    function close() {}
}
